### PR TITLE
Replace checkout script with submodules

### LINKED ISSUE #2141: Convert checkout script to git submodules
### What new functionality do you need?

Use git submodules instead of a checkout script. This will bring us in line with other repos and SOP.

### What are the requirements for the new functionality?

Git submodules for each of the components that `sorc/checkout.sh` current checks out, set to the corresponding hash. The options formerly used with checkout to limit the components checked out moved to the build all script.

### Acceptance Criteria

- [ ] `sorc/checkout.sh` replaced with appropriate git submodules
- [ ] `sorc/build_all.sh` updated to optionally build components based on switches instead of keying on directories, since all will be present
- [ ] CI updated to use submodules and not checkout.sh
- [ ] Documentation updated

### Suggest a solution (optional)

_No response_

### PR DESCRIPTION
**Pull request by WalterKolczynski-NOAA**
# Description
Replaces the checkout script with git submodules. This change brings workflow in line with other repos and git standards.

Due to these changes, the build system is updated slightly. First, the options that were formerly used with `checkout.sh` to reduce unneeded builds (`-g` for GSI and `-u` for UFSDA) are moved to `build_all.sh` instead. This is needed since build all will not be able to key off directories existing or not to determine which components to build. Second, the partial build system is removed. It was rarely, if ever, used, and parts of it were already ignored with the previous changes to base builds off whether directories exist. The manage externals file is also removed since that is also not being used.

Following these changes, clones/checkout will need to either be recursive, or `git submodule update` will need to be called separately to make sure components are properly checked out (non-exhaustive list):

```
## Checkout develop (method 1)
git clone --recursive https://github.com/NOAA-EMC/global-workflow destination

## Checkout develop (method 2)
git clone  --jobs 8 https://github.com/NOAA-EMC/global-workflow destination
cd destination
git submodule update --init --recursive --jobs 8

## Checkout existing branch some_branch (method 1)
git clone --recursive  --jobs 8 -b some_branch https://github.com/NOAA-EMC/global-workflow destination

## Checkout existing branch some_branch (method 2)
git clone https://github.com/NOAA-EMC/global-workflow destination
cd destination
git checkout --recurse-submodules some_branch

## Checkout existing branch some_branch (method 3)
git clone https://github.com/NOAA-EMC/global-workflow destination
cd destination
git checkout some_branch
git submodule update --init --recursive  --jobs 8

## Checkout new branch new_branch
git clone https://github.com/NOAA-EMC/global-workflow destination
cd destination
git checkout -b new_branch
git submodule update --init --recursive  --jobs 8
```

The git ignore list had to be updated to use explicit paths as the wildcard paths pick up the ones that are now submodules.

UFS creates files during the build process that are not in its ignore list, so that submodule ignores the "dirty" state to avoid confusing users running `git status`.

CI and documentation are also updated to reflect the changes.

Resolves #2141 

# Type of change
- Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? YES
- Does this change require a documentation update? YES

# How has this been tested?
- Clone and build on Orion

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

**Comment by WalterKolczynski-NOAA** [comment]
@TerrenceMcGuinness-NOAA we'll need you to test the new CI works properly. With this type of change, I don't think there is much use in sending this through the normal CI channel; they'll have to be updated simultaneously.

File: docs/source/clone.rst [hunk 1]
```diff
@@ -9,41 +9,40 @@ Quick Instructions
 Quick clone/build/link instructions (more detailed instructions below).
 
 .. note::
-   Here we are making the assumption that you are using the workflow to run an experiment and so are working from the authoritative repository. If you are using a development branch then follow the instructions in :doc:`development.rst`. Once you do that you can follow the instructions here with the only difference being the repository/fork you are cloning from.
+   Here we are making the assumption that you are using the workflow to run an experiment and so are working from the authoritative repository . If you are using a development branch then follow the instructions in :doc:`development.rst`. Once you do that you can follow the instructions here with the only difference being the repository/fork you are cloning from.
```
Comment (KateFriedman-NOAA) [inline docs/source/clone.rst]: Remove the errant space that snuck in.

**Review by KateFriedman-NOAA — APPROVED** [review]
Looks great, thanks for working on this @WalterKolczynski-NOAA ! I left one tiny comment to remove an errant space. No other comments.

**Review by DavidHuber-NOAA — CHANGES_REQUESTED** [review]
Looks pretty good.  I ran a checkout/build on Orion and everything worked as expected.  The only thing I see is that the link script will create dead links to anything that isn't built.  Perhaps the check for GDAS should be `[[ -d "${HOMEgfs}/sorc/gdas.cd/bin" ]]` and for the GSI, GSI-monitor, and GSI-utils: `[[ -d "${HOMEgfs}/sorc/<gsi_dir>/install" ]]`.  Otherwise, this looks really great!

File: docs/source/clone.rst (hunk 1 above)
Comment (WalterKolczynski-NOAA) [reply docs/source/clone.rst]: Fixed

**Comment by HenryRWinterbottom** [comment]
LGTM. Thanks @WalterKolczynski-NOAA.

**Comment by aerorahul** [comment]
I don't think it is this PR, but there are some dead links referring to GDASApp executables after `link_workflow.sh` is executed.  See the `red` ones in the image below:

Screenshot 2023-12-12 at 12 33 13 PM

@CoryMartin-NOAA @RussTreadon-NOAA Are these being built correctly in the GDASApp?  Are they being used in the global applications?

**Review by aerorahul — APPROVED** [review]
Looks good.
I don't see the issues @DavidHuber-NOAA reported.

**Comment by CoryMartin-NOAA** [comment]
yeah I think names changed, @aerorahul, we can fix it "soon". My priority this afternoon/tomorrow is to get GDASApp to also use submodules. I will fix this in the PR that adds the new submodule for GDASApp.

**Comment by aerorahul** [comment]
> yeah I think names changed, @aerorahul, we can fix it "soon". My priority this afternoon/tomorrow is to get GDASApp to also use submodules. I will fix this in the PR that adds the new submodule for GDASApp.

no rush.  Just noting something has changed.

**Comment by DavidHuber-NOAA** [comment]
@aerorahul I cloned the branch, built just the core programs (i.e. no DA), and then linked the workflow as follows:

```
git clone git@github.com:WalterKolczynski-NOAA/global-workflow -b feature/submodules --recursive
cd global-workflow/sorc
./build_all.sh
./link_workflow.sh
```

This resulted in several dead links in the exec directory as shown below.  The `link_workflow.sh` script only looks for the presence of the head of the directories (e.g. `gdas.cd`) and links based on that rather than what was built.  Since only a subset of users will build the GDASApp, I would prefer that those links only be created if a user is planning on using that system.  Similarly, some users may not want to build DA at all and only run forecasts or build the GDASApp without the GSI.
Screenshot 2023-12-12 135046

**Comment by WalterKolczynski-NOAA** [comment]
> @aerorahul I cloned the branch, built just the core programs (i.e. no DA), and then linked the workflow as follows:
> 
> ```
> git clone git@github.com:WalterKolczynski-NOAA/global-workflow -b feature/submodules --recursive
> cd global-workflow/sorc
> ./build_all.sh
> ./link_workflow.sh
> ```
> 
> This resulted in several dead links in the exec directory as shown below. The `link_workflow.sh` script only looks for the presence of the head of the directories (e.g. `gdas.cd`) and links based on that rather than what was built. Since only a subset of users will build the GDASApp, I would prefer that those links only be created if a user is planning on using that system. Similarly, some users may not want to build DA at all and only run forecasts or build the GDASApp without the GSI. Screenshot 2023-12-12 135046

Yeah, I'm planning on addressing this, I have just been stuck in meetings all day.

**Comment by WalterKolczynski-NOAA** [comment]
@DavidHuber-NOAA Broken links when skipping UFSDA/GSI build should be gone now.

**Review by DavidHuber-NOAA — APPROVED** [review]
Looks good, thanks @WalterKolczynski-NOAA!

**Comment by DavidHuber-NOAA** [comment]
This is beyond the scope of this PR, but I noted that there are some other broken links in the workflow, noted below

image

**Comment by DavidHuber-NOAA** [comment]
I looked into the CI test failure.  It was another timeout of `radmon_time.x`.  I noted this one other time.  Perhaps there is an intermittent hang?

**Comment by aerorahul** [comment]
> This is beyond the scope of this PR, but I noted that there are some other broken links in the workflow, noted below
> 
> image

The `global_angupdate` dead link is because of someone named @aerorahul might have tried to not duplicate them.  Since the packages of gsi and gsi-utils are separated, it might need duplication.  On the other hand, global_angupdate has been long deprecated and remains a relic in gsi-utils repository.

**Comment by TerrenceMcGuinness-NOAA** [comment]
```
mterry (Orion-login-1) global-workflow (feature/submodule) $ clear
mterry (Orion-login-1) global-workflow (feature/submodule) $ source versions/build.spack.ver
mterry (Orion-login-1) global-workflow (feature/submodule) $ source versions/run.orion.ver
mterry (Orion-login-1) global-workflow (feature/submodule) $ module load module_base.orion
mterry (Orion-login-1) global-workflow (feature/submodule) $ ldd exec/radmon_time.x | head
	linux-vdso.so.1 =>  (0x00007ffccff8f000)
	libnetcdff.so.7 => /work/noaa/epic/role-epic/spack-stack/orion/spack-stack-1.5.1/envs/unified-env/install/intel/2022.0.2/netcdf-fortran-4.6.0-i2wplrg/lib/libnetcdff.so.7 (0x00007fc2ddc79000)
	libnetcdf.so.19 => /work/noaa/epic/role-epic/spack-stack/orion/spack-stack-1.5.1/envs/unified-env/install/intel/2022.0.2/netcdf-c-4.9.2-j2wl3gb/lib/libnetcdf.so.19 (0x00007fc2dd7fb000)
	libm.so.6 => /lib64/libm.so.6 (0x00007fc2dd4f9000)
	libmpifort.so.12 => /apps/intel-2022.1.2/intel-2022.1.2/mpi/2021.5.1/lib/libmpifort.so.12 (0x00007fc2dd145000)
	libmpi.so.12 => /apps/intel-2022.1.2/intel-2022.1.2/mpi/2021.5.1/lib/release/libmpi.so.12 (0x00007fc2db910000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fc2db70c000)
	librt.so.1 => /lib64/librt.so.1 (0x00007fc2db504000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fc2db2e8000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fc2daf1b000)
mterry (Orion-login-1) global-workflow (feature/submodule) $ time exec/radmon_time.x --help
```

Looks line `radmon_time.x` is hanging --- no, I see it is waiting for input, never-mind. Waiting for other's input.

**Comment by WalterKolczynski-NOAA** [comment]
Orion is clearly unstable again, as it is failing on a different job each time.

**Comment by WalterKolczynski-NOAA** [comment]
Remaining experiment is only waiting for archive and cleanup, but is stuck queued because of the imminent maintenance downtime. Going to set this to passed manually and merge.

**Pull request merged**